features/310-prerender-gear/prerender.md

```diff
@@ -0,0 +1,105 @@
+
+To allow scripts to run to its completion or to allow DOM to transform to a specific state, prerender waits for a configurable period of time before capturing the resultant DOM. The resultant DOM will have all of its script tags removed, and a `<base>` tag is inserted to the DOM appropriately to fix relative links.
+
+Response headers are then modified. Specifically, `content-length` and `content-encoding` will be removed.
+
+Finally, the modified DOM and headers are put into prerendered html cache using the forwarded url as keys before being sent back to gateway. A configurable expiry time is also applied to the cache such that they expire automatically.
+
+#### Redirect
+
+Note that since the forwarded url is used directly as the cache key, even if the forwarded url in a future request is the same as any of the redirected urls prerender has encountered during navigation, the request is a cache miss. For example:
+
+1. gateway requests url `http://a.com`
+2. prerender sees a caches miss, proceed to navigates to `http://a.com`
+3. prerender is redirected to `http://b.com`
+4. prerender captures the DOM of `http://b.com`, then stores it with key `http://a.com` and serves it
+5. gateway requests url `http://b.com`
+6. prerender sees a cache miss (it only has cache for `http://a.com`), proceeds to navigates to `http://b.com`
+7. prerender captures the DOM of `http://b.com` (again), then stores it with key `http://b.com` and serves it
+
+Notice how in (6) prerender consider `http://b.com` a cache miss even though it has been redirected to it.
+
+### Invalidating cache
+
+Since prerender supports some configuration about cache, when a configuration is changed such that according to the new configuration old cache is no longer valid, it is necessary to manually reset prerendered html cache. Prerender supports invalidation via a `origin` and `path` protocol.
+
+## API
+
+### Configuration
+
+`seo-prerender`: Boolean. `true` to enable prerender, `false` to disable.
+
+`seo-prerender-expiry`: Time. The amount of time a prerendrered html is kept in cache.
+
+`seo-prerender-capture-wait`: Time. The amount of time prerender waits before capturing DOM such that javascript, etc can have time to run.
+
+`seo-prerender-user-agent`: Comma separated strings. `user-agent` that is considered a search engine.
+
+### Endpoint
+
+#### Prerender
+
+Path:
+
+`/`
+
+Query parameters:
+
+- `u`: The url to render. A url with slash at the end and a url without slash at the end are cached separately.
+- `expiry`: Corresponds to `seo-prerender-expiry` in configuration, with the time specified in seconds. i.e. To have an expiry of one hour one should pass `3600`. Optional.
```

Comment:
What is the default value if it is not given?

Reply:
Would add back.